src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeActions.ConvertLinq
+{
+    public class ConvertForEachToLinqQueryTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+            => new CodeAnalysis.CSharp.ConvertLinq.ConvertForEachToLinqQuery.CSharpConvertForEachToLinqQueryProvider();
+
+        #region Query Expressions
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task QueryForForWhere()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class Query
+{
+    public IEnumerable<int> void Main(string[] args)
+    {
+        List<int> c1 = new List<int>{1, 2, 3, 4, 5, 7};
+        List<int> c2 = new List<int>{10, 30, 40, 50, 60, 70};
+        [|foreach (var x1 in c1)
+        {
+            foreach (var x2 in c2)
+            {
+                if (object.Equals(x1, x2 / 10))
+                {
+                    yield return x1 + x2;
+                }
+            }
+        }|]
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class Query
+{
+    public IEnumerable<int> void Main(string[] args)
+    {
+        List<int> c1 = new List<int>{1, 2, 3, 4, 5, 7};
+        List<int> c2 = new List<int>{10, 30, 40, 50, 60, 70};
+        return from x1 in c1
+               from x2 in c2
+               where object.Equals(x1, x2 / 10)
+               select x1 + x2;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task QueryForVarForWhere()
+        {
+            string source = @"
+using System.Linq;
+
+class C
+{
+    IEnumerable<int> M()
+    {
+        [|foreach (var num in new int[] { 1, 2 })
+        {
```

Comment:
what is error case behavior? like n1 is already define outside and etc?

Reply:
If there is an error case, we do not provide a code action. If there is an assignment instead of declaration, we stop conversion at the point.

---
In reply to: [195852622](https://github.com/dotnet/roslyn/pull/27130#discussion_r195852622) [](ancestors = 195852622)